Handlers/AdvancedSKUManagementHandler.cs

```diff
@@ -0,0 +1,101 @@
+            Lazy<ISKUGenerationServices> SKUGenerationServices) {
+
+            _orchardServices = orchardServices;
+            _contentManager = contentManager;
+
+            _SKUUniquenessExceptionProviders = SKUUniquenessExceptionProviders;
+
+            _SKUGenerationServices = SKUGenerationServices;
+
+            T = NullLocalizer.Instance;
+
+            OnUpdated<ProductPart>((ctx, part) => CreateSku(part));
+            OnCreated<ProductPart>((ctx, part) => {
+                if (part.ContentItem.VersionRecord == null) {
+                    CreateSku(part);
+                }
+            });
+        }
+
+        public Localizer T { get; set; }
+
+        protected override void UpdateEditorShape(UpdateEditorContext context) {
+            base.UpdateEditorShape(context);
+            //Apply this only for ProductParts
+            if (context.ContentItem.Parts.Any(part => part is ProductPart)) {
+                var productPart = context.ContentItem.As<ProductPart>();
+                var settings = _orchardServices.WorkContext.CurrentSite.As<AdvancedSKUsSiteSettingPart>();
+                if (settings.RequireUniqueSKU) {
+                    var sameSKUProductIds = _contentManager.Query<ProductPart, ProductPartRecord>(VersionOptions.Latest)
+                        .Where(ppr => ppr.Id != productPart.Record.Id && ppr.Sku == productPart.Sku)
+                        .List().Select(pp => pp.Id);
+                    //Handle exceptions to uniqueness of SKUs
+                    if (_SKUUniquenessExceptionProviders.Any()) {
+                        var exceptionIds = _SKUUniquenessExceptionProviders.SelectMany(provider => provider.GetIdsOfValidSKUDuplicates(productPart));
+                        sameSKUProductIds = sameSKUProductIds.Where(ppid => !exceptionIds.Contains(ppid));
+                    }
+                    if (sameSKUProductIds.Any()) {
+                        context.Updater.AddModelError("", T("The SKU must be unique."));
+                    }
+                }
+            }
+        }
+
+        private void CreateSku(ProductPart part) {
+            if (_SKUGenerationServices.Value.GetSettings().GenerateSKUAutomatically) {
+                ProcessSku(part);
+            }
+        }
+
+        private void ProcessSku (ProductPart part) {
+            if (string.IsNullOrWhiteSpace(part.Sku) ||
+                _SKUGenerationServices.Value.GetSettings().AllowCustomPattern) { //AllowCustomPattern is handled inside GenerateSku
+                //generate a new sku
+                part.Sku = _SKUGenerationServices.Value.GenerateSku(part);
+            }
+
+            //if the SKU is empty, generate a new one
+            if (string.IsNullOrWhiteSpace(part.Sku)) {
+                _SKUGenerationServices.Value.ProcessSku(part);
+                _orchardServices.Notifier.Warning(T("A new SKU has been generated: \"{0}\"", part.Sku));
```

Comment:
You are probably right, I will change that. Note however that in the corresponding code in the `ProcessAlias` method of `AutoroutePartHandler` it uses a warning.